### PR TITLE
Add solver logic for minesweeper

### PR DESCRIPTION
**Pull request by SeanMASilva**
This is untested code currently. I've never coded in Java so I need a review for coding patterns etc. currently.

### CODE REVIEW

**Comment by SeanMASilva** [comment]
The solver logic is based from here https://github.com/ghewgill/puzzles/blob/master/mines.c which has it looks like MIT license, where should I add the licence (big comment at the top of MSBoardSolver.java)?

**Comment by mitchej123** [comment]
If you're using mit licensed code you need to include the copyright header per the license

**Comment by SeanMASilva** [comment]
> If you're using mit licensed code you need to include the copyright header per the license

I'm asking where the header should go, because the whole project won't now become MIT licensed. Do I just add it at the top of `MSBoardSolver.java` and `MineSets.java` in a block comment, or is there somewhere else it should go?

**Comment by mitchej123** [comment]
In that case you'd generally put the current license of your changes in the file, and include the original (with indication original license)

(paraphrased)
Modifications made available under the <LGPL/appropriate license>

Original license
<blah>

**Comment by SeanMASilva** [comment]
Tested it now, seems to be working now

File: src/main/java/ru/timeconqueror/lootgames/common/config/ConfigMS.java [hunk 1]
```diff
@@ -54,6 +55,13 @@ public void init() {
                 1,
                 Integer.MAX_VALUE,
                 "It represents the number of attempts the player has to beat the game successfully.");
+        boardLogic = config.getInt(
```
Comment (ABKQPO) [inline src/main/java/ru/timeconqueror/lootgames/common/config/ConfigMS.java:58]: why not use boolean?

File: src/main/java/ru/timeconqueror/lootgames/minigame/minesweeper/MineSets.java [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+     */
+    public MineSets() {
+        this.sets = new TreeSet<>();
+        this.todo = new TreeSet<>();
+    }
+
+    public void addSet(int set) {
+        int mask = LogicMineSet.getSetMask(set);
+        if (mask == 0) {
+            System.out.println("Cannot add set with invalid mask. " + LogicMineSet.toString(set));
```
Comment (ABKQPO) [inline src/main/java/ru/timeconqueror/lootgames/minigame/minesweeper/MineSets.java]: use LootGames.LOGGER

File: src/main/java/ru/timeconqueror/lootgames/minigame/minesweeper/MSBoardSolver.java [hunk 3]
```diff
@@ -0,0 +1,662 @@
...
+                    for (int bombs : p.descendingKeySet()) {
+                        if (bombs + minBrute > maxPlace) {
+                            highKey = bombs;
+                        }
+                    }
+                    pruned = pruned || lowKey != -1 || highKey != -1;
+                    if (lowKey != -1) p.headMap(lowKey).clear();
+                    if (highKey != -1) p.tailMap(highKey).clear();
+                    minBrute += p.firstKey();
+                    maxBrute += p.lastKey();
```
Comment (ABKQPO) [inline src/main/java/ru/timeconqueror/lootgames/minigame/minesweeper/MSBoardSolver.java:530]: `headMap(lowKey)` excludes `lowKey`, but it’s already invalid here, so it should be removed inclusively (e.g. `headMap(lowKey, true).clear()`) to avoid incorrect pruning results; consider adding a regression test for this boundary.

**Review by ABKQPO — CHANGES_REQUESTED** [review]
@SeanMASilva

File: src/main/java/ru/timeconqueror/lootgames/common/config/ConfigMS.java (hunk 1 above)
Comment (SeanMASilva) [reply src/main/java/ru/timeconqueror/lootgames/common/config/ConfigMS.java:58]: In the AOI the original discussion was about an interactive solver that would reveal a square, iff a guess was required. Tooj and Four? also talked about how guessing optimally was a fun part of the game. So I was planning ahead for if I wanted to write up an interactive solver / weaker solver (one that removed only 50/50s)

Should I just keep it as boolean and change it to int / string enum when I get around to writing different solvers?

Comment (ABKQPO) [reply src/main/java/ru/timeconqueror/lootgames/common/config/ConfigMS.java:58]: if this is intended by design, no changes are needed

File: src/main/java/ru/timeconqueror/lootgames/minigame/minesweeper/LogicMineSet.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        int x = LogicMineSet.getSetX(set);
+        int y = LogicMineSet.getSetY(set);
+        int mask = LogicMineSet.getSetMask(set);
+        int bombs = LogicMineSet.getSetBombs(set);
+        return "Mineset {x: " + x + " y: " + y + " mask: " + Integer.toBinaryString(mask) + " bombs: " + bombs + "}";
+    }
+
+    // n choose k
+    public static Iterable<List<Integer>> permutations(int n, int k) {
+        return () -> new java.util.Iterator<>() {
```
Comment (ABKQPO) [inline src/main/java/ru/timeconqueror/lootgames/minigame/minesweeper/LogicMineSet.java]: same

File: src/main/java/ru/timeconqueror/lootgames/minigame/minesweeper/LogicMineSet.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    public static int Set(int x, int y, int mask, int count) {
+        return LogicMineSet.setSetX((byte) x) | LogicMineSet.setSetY((byte) y)
+                | LogicMineSet.setSetMask(mask)
+                | LogicMineSet.setSetMines((byte) count);
+    }
+
+    public static Iterable<Integer> iterateSetMask(int set) {
+        return () -> new java.util.Iterator<>() {
```
Comment (ABKQPO) [inline src/main/java/ru/timeconqueror/lootgames/minigame/minesweeper/LogicMineSet.java]: please avoid fully qualified type references here and use imports instead

File: src/main/java/ru/timeconqueror/lootgames/minigame/minesweeper/MSBoardSolver.java [hunk 6]
```diff
@@ -0,0 +1,680 @@
...
+            changedOtherCells = this.board.perturbBombLocations(cellsToChange, this.boardKnowledge);
+        }
+
+        // If we have to backtrack then the mask's of the sets will not correlate with the
+        // boundary of known and unknown tiles, we need to rebuild all the sets.
+        // Initially clear the current sets to make the perturb applications more efficient
+        // (SetOverlap is quicker on empty tree)
+        if (backtrackDepth != initialBacktrackDepth) this.setStore.clear();
+
+        // The board returns a list of negative numbers if it filled cellsToChange with bombs
```
Comment (SeanMASilva) [inline src/main/java/ru/timeconqueror/lootgames/minigame/minesweeper/MSBoardSolver.java:604]: This is outdated

File: src/main/java/ru/timeconqueror/lootgames/minigame/minesweeper/MSBoardSolver.java [hunk 7]
```diff
@@ -0,0 +1,680 @@
...
+        // Apply the perturb's changes to the solver's knowledge
+        LootGames.LOGGER.trace("Applying {} perturbations", changedOtherCells.size());
+        for (BombPerturbation bp : changedOtherCells) {
+            this.applyPerturb(bp);
+        }
+        board.applyPerturbations(changedOtherCells);
+        LootGames.LOGGER.trace(boardKnowledgeToString());
```
Comment (SeanMASilva) [inline src/main/java/ru/timeconqueror/lootgames/minigame/minesweeper/MSBoardSolver.java:611]: Should I replace this with a formatted string (and other places), or don't bother, because we'll remove the traces once tested in dailies?

Comment (ABKQPO) [reply src/main/java/ru/timeconqueror/lootgames/minigame/minesweeper/MSBoardSolver.java:611]: looks like it is not needed

**Review by ABKQPO — APPROVED** [review]
overall, it looks pretty good
